rust_src/src/eval.rs

```diff
@@ -1,91 +1,100 @@
-//! Generic Lisp eval functions and macros.
+//! Generic Lisp eval functions
 
-/// Macro to generate an error with a list from any number of arguments.
-/// Replaces xsignal0, etc. in the C layer.
-///
-/// Like `Fsignal`, but never returns. Can be used for any error
-/// except `Qquit`, which can return from `Fsignal`. See the elisp docstring
-/// for `signal` for an explanation of the arguments.
-macro_rules! xsignal {
-    ($symbol:expr) => {
-        unsafe {
-            ::remacs_sys::Fsignal($symbol, ::remacs_sys::Qnil);
-        }
-    };
-    ($symbol:expr, $($tt:tt)+) => {
-        unsafe {
-            ::remacs_sys::Fsignal($symbol, list!($($tt)+).to_raw());
-        }
-    };
-}
+use remacs_macros::lisp_fn;
+use remacs_sys::Qnil;
+use remacs_sys::eval_sub;
 
-/// Macro to call Lisp functions with any number of arguments.
-/// Replaces CALLN, call1, etc. in the C layer.
-macro_rules! call {
-    ($func:expr, $($arg:expr),*) => {{
-        let mut argsarray = [$func.to_raw(), $($arg.to_raw()),*];
-        unsafe {
-            LispObject::from_raw(
-                ::remacs_sys::Ffuncall(argsarray.len() as ::libc::ptrdiff_t, argsarray.as_mut_ptr())
-            )
-        }
-    }}
-}
+use lisp::LispObject;
+use lisp::defsubr;
 
-macro_rules! call_raw {
-    ($func:expr, $($arg:expr),*) => {{
-        let mut argsarray = [$func, $($arg),*];
-        unsafe {
-            LispObject::from_raw(
-                ::remacs_sys::Ffuncall(argsarray.len() as ::libc::ptrdiff_t, argsarray.as_mut_ptr())
-            )
+/// Eval args until one of them yields non-nil, then return that value.
+/// The remaining args are not evalled at all.
+/// If all args return nil, return nil.
+/// usage: (or CONDITIONS...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn or(args: LispObject) -> LispObject {
+    let mut val = Qnil;
+
+    for elt in args.iter_cars_safe() {
+        val = unsafe { eval_sub(elt.to_raw()) };
+        if val != Qnil {
+            break;
         }
-    }}
+    }
+
+    LispObject::from_raw(val)
 }
 
-macro_rules! message_with_string {
-    ($str:expr, $obj:expr, $should_log:expr) => {
-        unsafe {
-            ::remacs_sys::message_with_string($str.as_ptr() as *const ::libc::c_char,
-                                              $obj.to_raw(),
-                                              $should_log);
+/// Eval args until one of them yields nil, then return nil.
+/// The remaining args are not evalled at all.
+/// If no arg yields nil, return the last arg's value.
+/// usage: (and CONDITIONS...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn and(args: LispObject) -> LispObject {
+    let mut val = Qnil;
```

Comment:
The default value should be Qt here. Would you mind adding a test too? `(and) => t`

Reply:
Good catch on the `t` that is a little counter intuitive.